### PR TITLE
web3.js: add getSignatureStatuses to confirm transaction

### PR DESCRIPTION
**Pull request by abrzezinski94**
#### Problem

If you call confirmTransaction for a signature that has already been processed, the signatureSubscribe subscription will never fire a notification. This will certainly mean that confirmTransaction() will throw an expiry error for a transaction that is, in actual fact, confirmed.

https://github.com/solana-labs/solana/issues/25955

#### Summary of Changes

add getSignatureStatuses

### CODE REVIEW

**Review by steveluscher — CHANGES_REQUESTED** [review]
Hey! Thanks for this PR; I really appreciate the help to make transaction confirmation more robust.

I have ideas to make this _way_ simpler.

You've implemented a _polling_ strategy here, and have done so without removing the existing subscription strategy. This means that we now have two competing strategies, where what we could do instead is to make them _compliment_ each other.

In the original issue, I wrote:

> 1. Before subscribing or honouring expiry signals, check the signature status using the getSignatureStatuses method.
> 2. If the signature has not been confirmed and you have not received an expiry signal, subscribe for status updates.
> 3. Race the expiry strategy and the confirmation strategy.

What I was getting at there is that – between the time you send the transaction and you subscribe for the signature – you can _miss_ the subscription notification.

The solution for this should be simple:

1. Subscribe first.
2. Immediately after the subscription is live, send a _single_ `getSignatureStatuses` to rule out the possibility of having missed a prior signature change.

This should make this PR almost literally one line, and should eliminate the inefficiency of polling!

Want to take a crack at that?

**Comment by abrzezinski94** [comment]
> Hey! Thanks for this PR; I really appreciate the help to make transaction confirmation more robust.
> 
> I have ideas to make this _way_ simpler.
> 
> You've implemented a _polling_ strategy here, and have done so without removing the existing subscription strategy. This means that we now have two competing strategies, where what we could do instead is to make them _compliment_ each other.
> 
> In the original issue, I wrote:
> 
> > 1. Before subscribing or honouring expiry signals, check the signature status using the getSignatureStatuses method.
> > 2. If the signature has not been confirmed and you have not received an expiry signal, subscribe for status updates.
> > 3. Race the expiry strategy and the confirmation strategy.
> 
> What I was getting at there is that – between the time you send the transaction and you subscribe for the signature – you can _miss_ the subscription notification.
> 
> The solution for this should be simple:
> 
> 1. Subscribe first.
> 2. Immediately after the subscription is live, send a _single_ `getSignatureStatuses` to rule out the possibility of having missed a prior signature change.
> 
> This should make this PR almost literally one line, and should eliminate the inefficiency of polling!
> 
> Want to take a crack at that?

Hi we use pooling like that in mango/gov ui and its quite nice. Websockets can sometimes fail and its good to have a backup method to confirm tx.

What do you think about adding option to config it for example:

a. websockets only
b. pooling only 
c. websockets + pooling (i would leave it as default method for reliability)

**Comment by steveluscher** [comment]
I would rather resolve the architectural decision here amongst ourselves, rather than to push it out to library users as a choice.

Some reasons I disfavor polling:
* It is _very_ low resolution. The maximum confirmation resolution that you can achieve is half the sampling frequency (as per [Nyquist](https://en.wikipedia.org/wiki/Nyquist_frequency)).
* It's wasteful of client connections, and RPC capacity; one subscription channel can report on _n_ signatures. Polling requires _n_ connections, _m_ times per second.

I think I would need some evidence that websockets are so unreliable – other than the _known_ problem of being able to ‘miss’ signatures that this PR aims to solve – as to warrant making every Solana application that uses `web3.js` start sending all extra network requests to RPCs. If there's evidence of the WS transport being that unreliable, we should fix that first.

**Review by steveluscher — CHANGES_REQUESTED** [review]
giphy-2

After those couple fixes we'll need these tests, and then we're done!

1. Test that `getSignatureStatuses` does not get fired off until `onSignature` has resolved.
2. Test that `getSignatureStatuses` does not get fired _at all_ if (somehow) `onSignature` resolves synchronously (who knows; that could get implemented in the future)
3. Test that transactions get confirmed if no subscription notifications get fired, but `getSignatureStatuses` returns a confirmation
4. Test that transactions still get confirmed if `getSignatureStatuses` returns nothing, but then you get a subscription notification.

You should be able to accomplish that without writing a ‘live’ test at all.

File: web3.js/src/connection.ts [hunk 1]
```diff
@@ -3295,6 +3295,24 @@ export class Connection {
           },
           subscriptionCommitment,
         );
+
+        (async () => {
```
[the remark this answers is not in the document]
Comment (abrzezinski94) [reply web3.js/src/connection.ts:3299]: isn't subscriptionId returned immediately from onSignature ?

[the remark this answers is not in the document]
Comment (abrzezinski94) [reply web3.js/src/connection.ts:3299]: at first glance maybe we could get hash like this.onSignature do it and observe this._subscriptionsByHash and when typeof this._subscriptionsByHash[hash] !== 'undefined' && state 'subscribed' fire gss and do it all in confirmation promise

[the remark this answers is not in the document]
Comment (abrzezinski94) [reply web3.js/src/connection.ts]: Ok i pushed some solution for you to look. still needs test if it fulfills all needs

**Review by steveluscher — DISMISSED** [review]
Let me take a crack at this.

**Comment by steveluscher** [comment]
@jordansexton, other than the fact that commented-out test doesn't work, what do you think of this (see my three commits in the history here).

**Comment by brianlong** [comment]
The solution proposed here looks good, and I do not see an undue burden on RPC nodes. The server load could go down if clients stop polling in a loop. Thanks for working on this!

**Pull request merged**